File: Backtracking/letter_phone_number_combination.cc
// Link: https://leetcode.com/problems/letter-combinations-of-a-phone-number/

class Solution 
{
    private:
        map<int, vector<char>> book;
        vector<string> ans;
        vector<int> no;
        int no_length = 0;
    
    public:
        void init_dict()
        {
            book.insert(make_pair(2, vector<char>({'a', 'b', 'c'})));
            book.insert(make_pair(3, vector<char>({'d', 'e', 'f'})));
            book.insert(make_pair(4, vector<char>({'g', 'h', 'i'})));
            book.insert(make_pair(5, vector<char>({'j', 'k', 'l'})));
            book.insert(make_pair(6, vector<char>({'m', 'n', 'o'})));
            book.insert(make_pair(7, vector<char>({'p', 'q', 'r', 's'})));
            book.insert(make_pair(8, vector<char>({'t', 'u', 'v'})));
            book.insert(make_pair(9, vector<char>({'w', 'x', 'y', 'z'})));
        }
    
        vector<string> letterCombinations(string digits)
        {
            no_length = digits.length();
            
            if(no_length == 0)
            {
                return ans;
            }
            
            init_dict();
            
            for(int i=0;i<no_length;i++)
            {
                no.push_back(digits[i]-48);
            }
            
            explore(0, "");
            return ans;
        }
    
        void explore(int idx, string prev)
        {   
            if(idx >= no_length)
            {
                ans.push_back(prev);
                return;
            }
            
            for(int i=0; i<book[no[idx]].size();i++)
            {
                explore(idx+1, prev+book[no[idx]][i]);
            }
        }
}; 

File: Backtracking/letter_phone_number_combination.py
# Link: https://leetcode.com/problems/letter-combinations-of-a-phone-number/

# Approach: Recursive Backtracking

class Solution(object):
    def init_digit_dic(self):
        self.dl = {}
        self.dl['2'] = ['a', 'b', 'c']
        self.dl['3'] = ['d', 'e', 'f']
        self.dl['4'] = ['g', 'h', 'i']
        self.dl['5'] = ['j', 'k', 'l']
        self.dl['6'] = ['m', 'n', 'o']
        self.dl['7'] = ['p', 'q', 'r', 's']
        self.dl['8'] = ['t', 'u', 'v']
        self.dl['9'] = ['w', 'x', 'y', 'z']
        
    def letter_combination(self, idx, digits, string):
        if idx >= len(digits):
            self.ans_list.append(string)
        else:
            for i in self.dl[digits[idx]]:
                self.letter_combination(idx+1, digits, string+i)
    
    def letterCombinations(self, digits):
        if digits == "":
            return []
        
        self.init_digit_dic()
        self.ans_list = []
        self.letter_combination(0, digits, '')
        return self.ans_list 
